### PR TITLE
Twitter handles aren't supposed to be case-sensitive

### PR DESCRIPTION
**Pull request by rowanseymour**
- Twitter URN normalization should make handles lowercase
- Removes Twitter URN duplicates
- Converts all existing ones to lowercase

### CODE REVIEW

File: temba/contacts/migrations/0029_norm_twitter_urns.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+            other_urns = [u for u in urns if u != last_used_urn]
+
+            # attach messages for the other URNs to the last used URN
+            for other_urn in other_urns:
+                num_msgs_moved = other_urn.msgs.update(contact_urn_id=last_used_urn.pk)
+                print " > Moved %d messages from handle %s to handle %s" % (num_msgs_moved, other_urn.path, last_used_urn.path)
+
+                # delete this URN
+                print " > Deleting URN #%d (%s)" % (other_urn.pk, other_urn.path)
+                other_urn.delete()
```
Comment (rowanseymour) [inline temba/contacts/migrations/0029_norm_twitter_urns.py]: One never likes deleting stuff.. but we enforce uniqueness on URNs so these duplicates have to go

Comment (norkans7) [inline temba/contacts/migrations/0029_norm_twitter_urns.py]: how does the foreign keys on Msg and Broadcast behave with this?

Comment (norkans7) [inline temba/contacts/migrations/0029_norm_twitter_urns.py]: Ohh just Broadcasts, msgs are well handled

Comment (rowanseymour) [inline temba/contacts/migrations/0029_norm_twitter_urns.py]: Good spot. URNs in broadcasts are fairly rare so didn't have problem on my local db, but have now added code to replace the dup URNs in broadcasts

**Comment by norkans7** [comment]
Looks good

**Comment by rowanseymour** [comment]
For deployment:

1. Stop Mage so it's no longer creating non-lowercase Twitter URNs
2. Push this
3. Update Mage with https://github.com/rapidpro/mage/pull/4

File: temba/contacts/migrations/0029_norm_twitter_urns.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
+# -*- coding: utf-8 -*-
+from __future__ import unicode_literals
+
+from collections import defaultdict
+from django.db import migrations
+from django.db.models import Func, F
+
+
+def get_chosen_one(urns):
```
Comment (nicpottier) [inline temba/contacts/migrations/0029_norm_twitter_urns.py]: "You were the chosen one!"

File: temba/contacts/migrations/0029_norm_twitter_urns.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        print " > Found %d Twitter handles with duplicates" % (len(duplicates))
+
+        for handle, urn_ids in duplicates.iteritems():
+            urns = list(ContactURN.objects.filter(pk__in=urn_ids).select_related('contact').prefetch_related('msgs'))
+
+            chosen_urn, other_urns = get_chosen_one(urns)
+
+            for other_urn in other_urns:
+                # attach messages for the other URNs to the last used URN
+                num_msgs_moved = other_urn.msgs.update(contact_urn_id=chosen_urn.pk)
```
Comment (nicpottier) [inline temba/contacts/migrations/0029_norm_twitter_urns.py]: Should we be moving contact over as well here?

Comment (nicpottier) [inline temba/contacts/migrations/0029_norm_twitter_urns.py]: Also, what about FlowRuns?

Comment (rowanseymour) [inline temba/contacts/migrations/0029_norm_twitter_urns.py]: Each of these duplicate URNs will typically have its own contact. So "moving" contacts would require merging them (groups, fields, runs etc) with the contact of the chosen one... and that sounds more complicated than I think is worth it for this small number of cases. No?

**Pull request merged**